### DISCUSSION TITLE
ARM OS Support Consolidation Poll

### DISCUSSION THREAD

**Discussion by shitwolfymakes** [automatic-ripping-machine/automatic-ripping-machine#698] [Polls]
Historically, supporting ARM across the current breadth of OS distros, versions, and Docker has been challenging. This has resulted in the manual install docs growing wildly out of date and the debian installer being left in a broken state. This is not ggod for the long-term health of the project and us as maintainers and developers

To be clear, we don't have the time (or hardware) to support all of this. The docker container is built on a [custom Ubuntu image](https://github.com/phusion/baseimage-docker) that is just now starting to add support for 22.04, and we haven't even begun to try figuring out what 22.04 changed that's causing ARM to break.

Additionally, Docker wasn't anywhere near the global force for portability that it is today when this project first started. This makes adding support for a new OS as simple as running a container and checking that everything works as expected.

So I propose the following:
1) ARM drops manual and automated installation tooling and support for all OSes other than Ubuntu 20.04 LTS (what the docker image and our dev stack is based on. If that changes then the installer support will change to reflect that).
2) Prioritize the usage of the Docker container over all other deployment types.

This will allow us to remediate bugs faster, keep the docs more up to date, and give us more time to work on features like #589 

Also, please feel to leave comments with your thoughts if that's possible on a poll

**Comment by hazenaj** [comment]
I'm a latecomer to the project but found that Docker works with a lot less pain than a failed setup on OMV. I have a lot less to worry/bother with in a container than if I screw up the OS, which is running other things as well (fileserver). While running native may be great on an appliance type of setup, Docker is awesome for a general server setup from what I've seen. Plus, running in Docker has given me an opportunity to learn 1 or 10 new things :-)

**Comment by shitwolfymakes** [comment]
Side note: our docker container includes support for the hardware acceleration that the current install doesn't. Part of the perks of precompiling a base image to our specifications!

**Comment by TheCrimsonLady** [comment]
Just dropping in my own two cents here.
While I am a big fan of native installs, especially for things like this, the problems you outlined and the convenience of Docker can not be ignored.

Thank you very much for this project!

**Comment by microtechno9000** [comment]
Having some consistent and known system states will make changes to ARM much easier to manage. At the moment testing requires multiple OS versions and distros, we can't test against everything so its going to miss some bugs or dependencies.

**Comment by 1337-server** [comment]
Docker is the way forward to make ARM as accessible as possible for as many as possible. It would remove a lot of development issues with various systems and would help keep consistency.
It also removes the sometimes tough barrier for entry to install for newer Linux users.

But for me personally, I would always keep a copy of the hardwired version running on my own server, as even the small performance hit doesn't seem worth it for me.

**Comment by IamTheCarl** [comment]
Personally, I would suggest that for people who want native installers, they should spin up their own repositories for that.
If people are interested in a native install, you can forward them to those repositories. The Docker container would be more of a reference implementation for people maintaining such packages to start from.

I think that would be a good way to distribute the work.

**Comment by notnarb** [comment]
I think focusing more on making "Docker" a first-class citizen would be a great idea.

When I first read [the GitHub readme requirements section](https://github.com/automatic-ripping-machine/automatic-ripping-machine/blob/a20f95ec9a73b3cb33457fcc2c2f0eadc4106f8d/README.md?plain=1#L54-L63) my heart sank because at first glance this seemed like an Ubuntu-specific project.

Fortunately this was not the case, but it required a little additional investigation to figure out how to specifically configure it[1]

---

I want to add that supporting Docker doesn't mean that specifically Docker is required to run the image. I ran the Docker image using [Podman](https://podman.io/) instead and that worked great.

The OCI image format used by Docker has grown beyond only being supported by Docker.

---

[1] FYI the [docker wiki page](https://github.com/automatic-ripping-machine/automatic-ripping-machine/wiki/docker) has a broken link to DockerHub because it links to https://hub.docker.com/repository/docker/automaticrippingmachine/automatic-ripping-machine which is inaccessible if you don't have permissions on the project.  This link referenced in the pinned GitHub issue appears to be correct.

**Comment by maccampb** [comment]
I’m a newcomer as well and have found great utility in ARM. While I am a Debian fan, I understand the workload involved. If this is moved to docker only, I would plan to copy a Debian setup and try to keep updates applied locally. My only request is to keep the software as system agnostic as possible to make it easier for others to port to Debian when needed. Regards,MACwww.ckmmconsulting.caVia iPhone 12.On Feb 3, 2023, at 11:23, James Carl ***@***.***> wrote:﻿
Personally, I would suggest that for people who want native installers, they should spin up their own repositories for that.
If people are interested in a native install, you can forward them to those repositories. The Docker container would be more of a reference implementation for people maintaining such packages to start from.
I think that would be a good way to distribute the work.

**Comment by shitwolfymakes** [comment]
Well, that's about as overwhelming an answer as I could've hoped for, so I'll get right on this

**Comment by flammableliquids** [comment]
I've been absent for a while due to pandemic and personal life issues, and I am excited to get back into this project.
I wanted to say thank you to you guys for trucking this along, faster and further than I thought anyone would be interest in (Specifically @shitwolfymakes @microtechno9000 @1337-server)

For me, the best part of this project, is that I could get old hardware, breathe life into them, doing something useful instead of becoming e-waste. The biggest speed boosters being a video card with NVENC support. Quicksync is not often supported on older CPU's.

I have seen many many posts on the forumsn of Plex and Nextcloud regarding problems with docker containers being unable to access hardware acceleration, specifically NVENC.

If I had realised that hardware acceleration (Quicksync and NVENC) is supported, I would have changed my vote to yes (Not that it really matters as yes has won : )   )

**Reply by shitwolfymakes** [reply]
Feel free to hop in the discord or start working on issues, the more the merrier!

**Comment by TigerC10** [comment]
Speaking as a new user to ARM, I can say those manual instructions caused me _much pain and suffering_.  I followed them to a T, and would get errors because files have moved around to new locations in the project.  I tracked down where they moved to and updating things as best I could, but there's still _a lot_ wrong with it.  I've had to re-image my linux box 4 times now because of various issues after installing ARM.

**Comment by trleith** [comment]
Also a newcomer -- a brand-newcomer. I've had an awful time and still haven't gotten things working quite correctly mainly, I think, because the first time I ever tried using Docker was three days ago and the only reason I "want" it in the first place is ARM, because ARM hasn't caught up officially with Ububntu 22.04.2 and I read [somewhere] that installing directly as a systemd thing is now deprecated. <wheww!> I get that Docker makes dev life easier, but holy smoke, not mine. If I wanted to figure out how to do a VM, I'd run one. FWIW I'm an old computer programmer, sometime sysadmin and all that. I've been around but I'm not good at it -- ideally I configure a machine and don't touch it for two or three years. It isn't like I get daily practice. But I have a reason to want ARM for now, so I thank you all for it and try to learn something.

The first thing one needs to know is how to get Docker running at all, and make the NVIDIA 1660 visible within it.  The documentation is lacking. "NVIDIA? I have no idea" is less than helpful. (Answer: NVIDIA provide a Docker package to sort of forward the GPU into the container, and (I think) you have to tell the Docker container --gpus). The next thing required is a recipe to map EVERY ARM/Docker directory onto the "real" Linux filesystem. For example, y'all keep talking about arm.yaml. Where the hell is it? "sudo find / -name arm.yaml -print" 
Answer:  /var/lib/docker/overlay2/c0306f23d4372a30ece768f86cb8aae4eb0b634298a33d01ff08c5e0dd8c9404/diff/opt/arm/setup/arm.yaml

WTH?? I thought the documentation said /opt/arm/config or some such.

This leads me down the Docker rabbit hole. Do I need " --mount type=bind,source=/opt/arm/config,target=/opt/arm/setup"?

and so on.

I think I'm getting there, but it's painful. Ripped a Blu Ray last night, but it didn't encode with the GPU because I ran http://svr2:8080/setup again and didn't go in and fix the settings again. And this because I don't know how to "fix" the Docker mounts without removing and replacing the ARM container. The MP4 container didn't get a correct name and Plex didn't identify the movie correctly, but at least it finished. It's all very frustrating. There are still things I don't understand and I've got 24 solid hours in it. It's a lot.

So yeah, do Docker if you must, but now the problem to solve is "get the naive noob past Docker". He's got to know how to find the log, run in debug mode, start, stop, remove, replace the container, all kinds of things, just to rip a first disk and start looking at what it does vs. what he needs. I don't know whether it's really any easier than native systemd X three or four flavors. This is not to mention HandBrakeCLI, abcde, MakeMKV, and arm itself.

**Comment by 1337-server** [comment]
> The first thing one needs to know is how to get Docker running at all, and make the NVIDIA 1660 visible within it

HW GPU encoding is currently not 100% working with NVENC and docker, we are working on this issue Here: https://github.com/automatic-ripping-machine/automatic-ripping-machine/issues/721
But it's more complex because we can't just run Nvidias docker runtime without causing other issues, so we are having to find ways around it.
NVENC does work on bare-metal Ubuntu though, I have tested it. As long as you build HandBrakeCLI from source.

I apologize for the lack of instructions for NVENC encoding, I only recently bought an Nvidia GPU, so it wasn't possibly for me to add instructions for it. A user did leave instructions, but I didn't update the WIKI with them. You can find them Here: https://github.com/automatic-ripping-machine/automatic-ripping-machine/issues/690

I have just updated the wiki for nvidia here: https://github.com/automatic-ripping-machine/automatic-ripping-machine/wiki/nvidia


> the only reason I "want" it in the first place is ARM, because ARM hasn't caught up officially with Ububntu 22.04.2

As for catching up with Ubuntu 22, it simply doesn't have one of the packages the script/installer needs.
`python3-psutil` is missing at last check a few weeks ago, so it's not possible for us to upgrade to 22 without dropping or finding alternative features.
**Installing arm on Ubuntu 22 has always been fully possible.** It was only the auto installer/scripts that are broken, you can find the instructions here https://github.com/automatic-ripping-machine/automatic-ripping-machine/issues/619#issuecomment-1415366785
Remove the 2 things I mention in that comment and run the installation script on 22 and everything on the ripper side of things will work, it's the UI that messes up.

I do personally fully understand that running on bare-metal much easier for those who are more familiar with Linux, but with the large number of flavors of Linux out there, it's much harder to support them all with the small number of developers we have.
Getting newer users to use docker is far easier, especially since they are more likely to have it installed or easily available on their system already.

Fully supporting running on bare-metal and every version of Linux takes away dev time from working on ARM itself, it's a fact. It slows down the dev team.
You can see this here https://github.com/automatic-ripping-machine/automatic-ripping-machine/pulse/monthly
This last month has been the biggest since active development restarted ~2 years ago.

**Reply by trleith** [reply]
I appreciate the reply and what y'all are trying to accomplish with Docker. I guess it'll help; you'll know better than I do.

FWIW, I did get HandBrakeCLI to start in the container and -z  showed  H.265 NVENC 2160p 4K as an available preset, so that much works. This is with the 525 driver and the NVIDIA Container Toolkit.

But there may be other issues:

trl@svr2:~$ HandBrakeCLI --version
[11:53:29] Compile-time hardening features are enabled
Cannot load libnvidia-encode.so.1
Cannot load libnvidia-encode.so.1
Cannot load libnvidia-encode.so.1
Cannot load libnvidia-encode.so.1
[11:53:29] hb_init: starting libhb thread
[11:53:29] thread 7fec481f3640 started ("libhb")
HandBrake 1.5.1

Maybe it has to do with "as long as you build HandBrakeCLI from source" which I have not done (yet). But maybe not: a quick web search turns up a TON on this particular symptom, with and without Docker.

Thanks again -- I'll have to see what I can accomplish, and try to take notes.

      t

**Reply by 1337-server** [reply]
We are trying to create a pre-built auto updated/generated version for everyone. If you are able to test it with us, it would be really helpful.

Image and instructions are here: https://github.com/automatic-ripping-machine/automatic-ripping-machine/issues/721#issuecomment-1439317446

**Reply by trleith** [reply]
Yeah, I can give it a go tomorrow.

**Reply by trleith** [reply]
Here's what I got today with automatic-ripping-machine:nvidia

```
[2023-11-02/24/23 19:11:47] DEBUG ARM: handbrake.handbrake_mkv Sending command: nice HandBrakeCLI -i /home/arm/media/raw/HARRY_CONNICK_JR_167726471215/D1_t00.mkv -o /home/arm/media/transcode/movies/HARRY_CONNICK_JR_167726471215/D1_t00.mkv --preset "H.265 NVENC 2160p 4K" --verbose >> /home/arm/logs/HARRY_CONNICK_JR_167726471215.log 2>&1
nice: ‘HandBrakeCLI’: No such file or directory
```

But I can shell into the container and run nice HandBrakeCLI -z, and I can ls -la all the paths that appear in the command line. I interpret the output to mean nice ran but couldn't launch HandBrakeCLI.

**Reply by trleith** [reply]
Oh.

There's more.

```
[2023-11-02/24/23 19:11:48] INFO ARM: utils.delete_raw_files Removing raw path - /home/arm/media/raw/HARRY_CONNICK_JR_167726471215
[2023-11-02/24/23 19:11:50] INFO ARM: utils.delete_raw_files Removing raw path - /home/arm/media/transcode/movies/HARRY_CONNICK_JR_167726471215
[2023-11-02/24/23 19:11:50] INFO ARM: utils.delete_raw_files Removing raw path - /home/arm/media/raw/HARRY_CONNICK_JR_167726471215
[2023-11-02/24/23 19:11:50] DEBUG ARM: utils.delete_raw_files No raw files found to delete in /home/arm/media/raw/HARRY_CONNICK_JR_167726471215 - [Errno 2] No such file or directory: '/home/arm/media/raw/HARRY_CONNICK_JR_167726471215'
[2023-11-02/24/23 19:11:50] DEBUG ARM: utils.database_adder Trying to add Notifications
[2023-11-02/24/23 19:11:50] DEBUG ARM: utils.database_adder successfully written Notifications to the database
[2023-11-02/24/23 19:11:50] INFO ARM: arm_ripper.rip_visual_media ************* ARM processing complete *************
```
If I shell into the container, I find 

```
root@b989d5001e3d:/home/arm# ls -la media/raw
total 12
drwxr-sr-x 3 arm arm 4096 Feb 24 19:11 .
drwxr-sr-x 8 arm arm 4096 Feb 24 18:16 ..
drwxr-sr-x 2 arm arm 4096 Feb 24 18:45 HARRY_CONNICK_JR
root@b989d5001e3d:/home/arm#

So there's an extra "Artist Name" directory level in media/raw (which is empty, I know not why). Maybe handbrake was starting after all, and complaining the input path it had been given was wrong.
```

**Reply by PriamX** [reply]
I entirely understand where the ARM maintainers are coming from here, in my professional life my department maintains large container swarms. Creating and developing ARM is one thing of itself, but creating and supporting a Docker container is entirely a different (yet related) skill set. In my professional world, the software developers are a different department and their product (the software) is handed to my department where we build, deploy, maintain and operate the customer-facing containers. In this poll, really what the maintainers are asking is: should we not only offer the ARM software and support, but should we also offer and support a container environment? Now mix in the desire to have HW encoding, and that complicates the container. HandBrakeCLI is used by ARM, and to use HW encoding with HandBrakeCLI you also need the proper drivers (seems Nvidia is discussed most here). And how is that best handled by the container? Do the ARM maintainers support hardware encoding? If so, this increases the complexity of the container, not only do they have to maintain a version of HandBrakeCLI that will use HW encoding for Intel and Nvidia and AMD, they would also need the respective GPU drivers exposed to the container. And this could get a bit fiddly with interdependencies. I feel the latter is going too far, perhaps biting off a bit too much...

My opinion is that if a Docker container is supported, then, for the ARM maintainers' sake, support only software encoding. However, like Fedora, like Ubuntu native, compile HandBrake for the ARM container such that if the GPU drivers are present, use them. Let setting up and configuring the nvidia runtime environment (for example) be the responsibility of the end-user (or installer). Certainly, provide some instructions on how to do this (for the ARM end-user, it could be no different than using Nvidia HW acceleration with Plex's docker container, for which there are existing examples). I would prefer the maintainers focus more on the ARM software development and feature set than spend time creating a Docker container that fits a multitude of use-cases. 

However, all that being said, I don't really know the current ARM-user demographic and may not entirely understand the specific objective(s) for the ARM Docker container.

**Comment by PriamX** [comment]
I'm not sure entirely what you mean by "support for native installation". I run ARM natively on Fedora. I run ARM in Docker. However, I may not be the demographic you're looking to, my Linux and development experience are extensive...I haven't had any trouble, nor won't have any trouble getting ARM to work under any Linux variety. If I were in your shoes I feel I would focus development on Docker, however, keep your installation procedure as documented as possible so folks like me can follow your installation instructions if we choose to run it on a native Linux...I'd consider that sufficient native support, and what I'm used to with open source projects. I do not feel it needs to be supported as much to have an apt-get or dnf repository, installing from source is fine by me, having a supported Docker container would be (IMHO) the best option for folks with less Linux experience to get into ARM.

**Comment by ghost** [comment]
For something like ARM, which is a collection of programs from different sources, a docker image would be a lot simpler to work with over a bare metal install. Instead of tracking dependencies you just need to have a docker script or compose file with like 5 things modified by the end user to suit their system.

Is it possible to consider using a smaller and more efficient/stable base like Alpine Linux, where you have more control of which versions of files are used? or would that have it's own dependency issues?

**Reply by microtechno9000** [reply]
Confused,  ARM is build on a docker image. The reason it is so large is MakeMKV, the build tools and GPU encoding support.

All those packages make it much larger. There is nothing stopping users modifying the base image and building their own ARM docker container rather than pulling one.

The devs have been focused on getting it to work for the majority, and making it easy to use. The thought is that if you are ripping a bluray, you can spare a few GB for a docker image. 

As always though, happy to take feedback from the community.

**Reply by ghost** [reply]
While size is a factor, one other factor is that Ubuntu is a more complicated distro compared to others, and I wonder if some of the difficulties you have encountered are exclusive to Ubuntu, and would not occur in a distro designed for docker use, such as Alpine Linux. Alpine removes many components that are not needed in a docker environment, and replaces systemd with a more minimalist init system. It's possible that an ARM container based on Alpine or something else would be easier to troubleshoot and maintain. Or it could not, if ARM is heavily reliant on Ubuntu-specific changes

**Reply by microtechno9000** [reply]
ARM is build on [Phusion](https://github.com/phusion/baseimage-docker), which is a modified Ubuntu systemd etc to work in a docker environment. 

Ubuntu was used as the bare metal scripts were based on an Ubuntu environment and when we migrated to docker a year or so ago, it was easier to keep things the same.

I am not familiar with Alpine and its benefits.

**Reply by DZaremba14600** [reply]
I build my own Docker image, and my changes include updating to the latest Phusion Ubuntu 22.04 image.  I'm running a host of Ubuntu 22.04 fully updated, and my docker image is ALSO running Ubuntu 22.04 (Phusion).  My changes are a very minor deviation from stock and I tried to document that as much as possible.

I can't say that I've fully exercised the ARM container with all its features and functions (especially CD ripping support), but I haven't had any issues that I'd suspect are related to the Phusion 'upgrade' from stock ARM.

I updated the container from 20.04 to 22.04 because the Intel QSV support seemed more recent and targeted to 22.04.

Jut an FYI.

**Reply by microtechno9000** [reply]
Welcome to raise a PR for review

**Reply by DZaremba14600** [reply]
Thanks for the invite.

I've never contributed via Git and am green/unfamiliar with the process.  What do I do?

I'm assuming I fork/clone/branch/edit/commit/push/ and then request a pull?

I'll watch a git beginners vid so do this properly.  I don't want to corrupt anything or violate any protocols or policies.

However, I upgraded to 22.04 in the container, run it on a 22.04 host, and have QSV enabled and running fine in the container. (I personally don't need to be cutting edge, but also don't want to get too far behind with regard to the Ubuntu version).

I've processed about 50 blurays with this and a handful of 4k blurays.

I'll follow up.

Thanks for the invite.

**Reply by DZaremba14600** [reply]
I haven't downloaded ARM in 5 or 6 weeks.  I was working with 2.8.1.  I see that 2.9 moved to the latest phusion-1.04.  Let me update my local copy and see about integrating my working QSV code which was just adding GPU support into the container.  Let me try this this in the next day or so.  If all is well, I'll try and see about a PR.

Thanks again.

**Reply by DZaremba14600** [reply]
I submitted a PR for the 'arm-dependencies' git project.  All my additions are in one file.

I noticed that there are a number of previous PR's.

I am currently using the stock ARM Docker Image with my changes and am processing probably 20 new disks and re-doing 50 or 60 disks to standardize some setting I've changed.  So far, no issues.

If you need anything besides the pull request, let me know.

Thanks.
